common/beerocks/bwl/shared/netlink_socket.cpp

```diff
@@ -0,0 +1,128 @@
+/* SPDX-License-Identifier: BSD-2-Clause-Patent
+ *
+ * Copyright (c) 2020 MaxLinear
+ *
+ * This code is subject to the terms of the BSD+Patent license.
+ * See LICENSE file for more details.
+ */
+
+#include "netlink_socket.h"
+
+#include <netlink/msg.h>
+#include <netlink/netlink.h>
+
+#include <easylogging++.h>
+
+namespace bwl {
+
+netlink_socket::netlink_socket(int protocol)
+    : m_nl_socket(nl_socket_alloc(), nl_socket_free), m_protocol(protocol)
+{
+    LOG_IF(!m_nl_socket, ERROR) << "Failed to allocate netlink socket!";
+}
+
+netlink_socket::~netlink_socket() {}
+
+bool netlink_socket::connect()
+{
+    // Connect the socket
+    if (nl_connect(m_nl_socket.get(), m_protocol) != 0) {
+        LOG(ERROR) << "Failed to connect netlink socket!";
+        return false;
+    }
+
+    return true;
+}
+
+void netlink_socket::close()
+{
+    // Terminate connection and close socket
+    nl_close(m_nl_socket.get());
+}
+
+bool netlink_socket::send_receive_msg(std::function<bool(struct nl_msg *msg)> msg_create,
+                                      std::function<bool(struct nl_msg *msg)> msg_handle)
+{
+    // The Netlink message to send
+    std::unique_ptr<nl_msg, void (*)(nl_msg *)> nl_message(nlmsg_alloc(), nlmsg_free);
+    LOG_IF(!nl_message, FATAL) << "Failed creating netlink message!";
+
+    // The Netlink callback set
+    std::unique_ptr<nl_cb, void (*)(nl_cb *)> nl_callback(nl_cb_alloc(NL_CB_DEFAULT), nl_cb_put);
+    LOG_IF(!nl_callback, FATAL) << "Failed creating netlink callback!";
+
+    // Termination flag for the loop that receives the response messages. Possible values are:
+    // error == 1: initial value, request message has not be sent yet
+    // error == 0: response has been successfully received
+    // error < 0: some error occurred while receiving response
+    // Final value is used to compute the result code of this method.
+    int error = 1;
+
+    // Create standard callbacks
+    static auto nl_err_cb = [](struct sockaddr_nl *nla, struct nlmsgerr *err, void *arg) -> int {
+        int *error = (int *)arg;
+        *error     = err->error;
+        LOG(ERROR) << "Failed to process netlink message! " << strerror(*error) << " (" << *error
+                   << ")";
+        return NL_STOP;
+    };
+
+    static auto nl_finish_cb = [](struct nl_msg *msg, void *arg) -> int {
+        int *error = (int *)arg;
+        *error     = 0;
+        return NL_SKIP;
```

Comment:
Probably it has the same effect returning NL_STOP than NL_SKIP. 

According to documentation,  NL_CB_FINISH gets called at the end of a multipart response message. According to source code, function nl_recvmsgs() continues parsing messages if NL_SKIP is received as a result of invoking callback and there are more messages to parse. So, if NL_CB_FINISH callback returns NL_SKIP and it's the last message in response, it has the same effect as if NL_STOP would have been returned.

However, extra messages after the multipart message would be ignored if NL_STOP were returned but not if NL_SKIP is returned. I don't know if this can be even possible, but just in case we should leave it as it is. All the examples I've checked in the Internet and original code in `base_wlan_hal_nl80211::send_nl80211_msg` return NL_SKIP too.